Maniramluitel2056/Customer-Churn-Analysis
# Merge CCA-21-final.version: Data Preprocessing, Feature Engineering, EDA Visualizations, and Conflict Resolutions

## Summary
This pull request merges the `CCA-21-final.version` branch into the `main` branch. It includes significant updates and enhancements to the Customer Churn Analysis project, focusing on data preprocessing, feature engineering, exploratory data analysis (EDA) visualizations, and the resolution of merge conflicts.

## Changes Implemented
### Data Preprocessing
- Handled missing values in the dataset using forward fill method.
- Encoded categorical variables using Label Encoding.
- Normalized numerical features using StandardScaler.

### Feature Engineering
- Introduced new features:
  - `Charges_Per_Tenure`: Monthly charges divided by (tenure + 1).
  - `TotalCharges`: Monthly charges multiplied by tenure.
  - `Contract_Type`: Encoded contract types (Month-to-month, One year, Two year).
  - `Payment_Method`: Encoded payment methods (Electronic check, Mailed check, Bank transfer, Credit card).

### Exploratory Data Analysis (EDA)
- Created various visualizations to understand the data better:
  - Distribution of churned and non-churned customers.
  - Histograms for numerical feature distributions.
  - Box plots for outlier detection.
  - Count plots for categorical feature analysis.
  - Correlation matrix for numerical features.
  - Pair plots with churn.
  - Group analysis by churn.

### Merge Conflict Resolutions
- Resolved conflicts in multiple files including:
  - `.vscode/settings.json`
  - `Data_Preparation/preprocessed_dataset/cleaned_dataset.csv`
  - `Data_Preparation/scaling_techniques/min_max_scaled_dataset.csv`
  - `Data_Preparation/scaling_techniques/standard_scaled_dataset.csv`
  - `data/interim/cleaned_dataset.csv`
  - `docs/data_loading_preprocessing.md` (deleted by us)
  - `notebooks/data_preprocessing/data_preprocessing.ipynb`
  - `scripts/data_preprocessing.py`
  - `utils/handle_missing_and_encode.py`
  - `utils/scaler.py`
- Ensured that our changes were preserved while integrating updates from the `main` branch.

### Additional Updates
- Updated documentation to reflect the changes in data processing and feature engineering.
- Ensured all scripts and notebooks are consistent and up-to-date with the latest changes.

## How to Test
1. Pull the latest changes from the `main` branch after the merge.
2. Run the `main.py` script to ensure all preprocessing, feature engineering, and EDA steps execute correctly.
3. Verify the generated visualizations and features in the output files.
4. Review the documentation for any updates or corrections needed.



This pull request represents a significant milestone in our project, integrating key features and improvements while resolving existing conflicts. Please review the changes and provide feedback.
